adds suggestion for collector set

xh3b4sd opened the pull request:
Towards https://github.com/giantswarm/giantswarm/issues/4462. Successor of https://github.com/giantswarm/operatorkit/pull/241.

kopiczko commented on collector/error.go:
```diff
@@ -0,0 +1,33 @@
+	}
+	if c == alreadyRegisteredError {
+		return true
+	}
```
IMO this condition should be the first to check. It's cheap to check and obvious we need it.

xh3b4sd replied:
Is that a rule or something? I don't mind actually. When this is important lets write it down.

kopiczko replied:
I don't think there is any rule regarding that. If I'm annoyed enough I'll write something.

kopiczko commented on collector/set.go:
```diff
@@ -0,0 +1,143 @@
+	s.logger.Log("level", "debug", "message", "collected metrics")
+}
+
+func (s *Set) CollectWithError(ch chan<- prometheus.Metric) error {
```
AFAIK it isn't part of the interface so can be private. And even named `collect`. This is a standard practice.

xh3b4sd replied:
This is part of the interface. Like, our interface our implementations should implement. `Collect(ch chan<- prometheus.Metric)` is the implementation of the Prometheus interface so the collector itself can be registered.

kopiczko commented on collector/spec.go:
```diff
@@ -0,0 +1,28 @@
+	// https://godoc.org/github.com/prometheus/client_golang/prometheus#Collector.
+	// The difference here is that this specific interface provides additional
+	// error handling capabilities.
+	CollectWithError(ch chan<- prometheus.Metric) error
```
Should be `Collect` and not return error? It isn't aligned with godoc.

xh3b4sd replied:
The godoc is outdated. We talked with Joe and we wanted our collectors to be able to return errors so we have nicer implementations there. This here is right IMO. We cannot call it `Collect` WITH have the error because the Prometheus interface already defines `Collect` WITHOUT error.

kopiczko replied:
Why does it have to implement Prometheus interface?

JosephSalisbury commented on collector/set.go:
```diff
@@ -0,0 +1,143 @@
+
+	err := s.CollectWithError(ch)
+	if err != nil {
+		s.logger.Log("level", "error", "message", "failed described metrics", "stack", fmt.Sprintf("%#v", microerror.Mask(err)))
```
`failed described metrics` - this should be `failed to collect metrics`, no?

xh3b4sd replied:
Sorted.

kopiczko commented on collector/set_test.go:
```diff
@@ -0,0 +1,104 @@
+
+//
+//
+//
```
What's that?

xh3b4sd commented:
Pawel and I talked. I simplified the interface and the `Set` type does not implement it anymore, but only the Prometheus one. I also removed the `Boot` function from the collector interface. PLease have another look.

xh3b4sd commented:
Note that the test I wrote is now rather useless. I used it to verify the boot process works fine, which it does, but the test does not really verify that now anymore because the collectors do not need to implement boot anymore.

kopiczko commented on collector/set.go:
```diff
@@ -0,0 +1,110 @@
+
+	if s.isBooted() {
+		return nil
+	}
```
Why not boot.Once?

xh3b4sd replied:
I wrote that in another comment. I think it was in the first proposal in `operatorkit`. I want the `Boot` interface to return an error. This is not really nicely possible with `sync.Once`.

kopiczko commented on collector/set.go:
```diff
@@ -0,0 +1,110 @@
+		err := c.Collect(ch)
+		if err != nil {
+			s.logger.Log("level", "error", "message", "failed collecting metrics", "stack", fmt.Sprintf("%#v", microerror.Mask(err)))
+			return
```
Should we return here? I think the loop should continue. It may be worthwhile to scrape other collectors.

JosephSalisbury replied:
I'd also like to see how we could have a generic error count

xh3b4sd replied:
I can remove the `return`. IMO the error count magic is a new feature we can address in a new PR.

xh3b4sd replied:
Removed the `return`.

kopiczko commented on collector/spec.go:
```diff
@@ -0,0 +1,21 @@
+	// The difference here is that this specific interface provides additional
+	// error handling capabilities.
+	Describe(ch chan<- *prometheus.Desc) error
+}
```
BTW it makes sense to me to put this interface in `spec.go`.

xh3b4sd replied:
I don't follow. This is already in `spec.go`.

JosephSalisbury commented on collector/set.go:
```diff
@@ -0,0 +1,110 @@
+	return s, nil
+}
+
+func (s *Set) Boot(ctx context.Context) error {
```
For my learning, why is this separated from `NewSet`? To have better control over _when_ it's registered?

xh3b4sd replied:
The constructor we usually call something like `New` should actually only initialize the structure and provide some useful implementation you can work with right away. All other procedures, which may be long running, should be done in a separate step. We call this `Boot`. There may be cases, like in testing situations, in which you only want the thing without it doing a lot of magic.

kopiczko replied:
How about:

```
func (s *Set) Boot(ctx context.Context) error {
	var err error
	s.bootOnce.Do(func() {
		err = s.boot(ctx)
	})

	return err
}

func (s *Set) boot(ctx context.Context) error {
	// ...
}
```

xh3b4sd replied:
With `sync.Once` we have a scope which we have to work around for the error. The same goes for goroutines. We do not need either of it here usually. Once being in such a scope the action might rather be asynchronous and cannot be returned anyway. I think it is easier and cleaner with the synchronous approach without scope.

kopiczko replied:
This approach is less error prone IMO. `sync.Once` is battle tested. Sync stuff is error prone in general. `sync.Once` is synchronous - this is known. Just saying.

xh3b4sd replied:
I actually took the implementation of `sync.Once` and used it here to get what I want. So this is just fine. We have the battle tested code.

JosephSalisbury commented on collector/set.go:
```diff
@@ -0,0 +1,108 @@
+	s.logger.Log("level", "debug", "message", "collecting metrics")
+
+	for _, c := range s.collectors {
+		err := c.Collect(ch)
```
Note for future: Sticking this in a waitgroup would be pretty cool

xh3b4sd replied:
Yeah we can totally iterate here and have a nice PR for that. No biggy. For now I am just glad we came up with a nice interface and can start using it. Parallelism and metrics and what not can easily be added.

JosephSalisbury replied:
This was my thinking.

JosephSalisbury commented:
When @kopiczko is happy :D
